runtimes/CMakeLists.txt

```diff
@@ -151,6 +151,9 @@ endif()
 # Avoid checking whether the compiler is working.
 set(LLVM_COMPILER_CHECKED ON)
 
+# This can be used to detect whether we're in the runtimes build.
+set(LLVM_RUNTIMES_BUILD ON)
+
 # Handle common options used by all runtimes.
 include(AddLLVM)
 include(HandleLLVMOptions)
```

Comment:
I actually don't know what we're picking up from `HandleLLVMOptions` but it all sounds like just dangerous to do in the context of the runtimes build. A naive build locally seems to work if we remove `include(HandleLLVMOptions)` entirely -- this is what I would try doing instead.

@petrhosek might know more.

Reply:
That was my first inclination, but I figured this would be the less destructive option. I'm assuming that many of these options are not required for projects that consume the LLVM static / shared libraries such as `openmp`, but I don't know the intention here strictly enough to make an accurate judgement there.

Reply:
I know for certain that I ran into a lot of issues back in the days where we used the `LLVM_ENABLE_PROJECTS` build caused by the rest of LLVM sneaking in flags via these kinds of CMake files. This is actively harmful. I think it's much better to remove it if we can, and normally our CI would tell us if we missed something.

So I'd support making the naive change that takes us closer to where we should be, and then figuring out where we went wrong in case it breaks something. You can always blame it on me :)

Reply:
It's been a few years since I last looked into this, but I tried removing it (since I agree that it can be actively harmful) and the issue I hit was that many of the runtimes builds were assuming variables and options set by `HandleLLVMOptions` because those were always present when building with `LLVM_ENABLE_PROJECTS`.

I'm not sure how much the situation has improved since then, I expect it won't be a big issue for libunwind, libc++abi, libc++ which no longer support `LLVM_ENABLE_PROJECTS`, but might still be an issue for compiler-rt and libc which do support building with `LLVM_ENABLE_PROJECTS` (although I hope that won't be the case for too long).

Probably the best path forward would be to try and remove it, see what fails and address each of those cases separately. The main issue is that many of those failures would be silent (e.g. a check that was passing before might be failing now but the build itself would still continue, except the build outputs might be different).

Reply:
We could add the `include` in compiler-rt and in libc if we expect those to rely on it perhaps? For libc++, libc++abi and libunwind, I am fairly confident that if we remove it and our CI doesn't break, we're okay. It might break something, but if that's the case and if it's actively being used then we have pretty good systems in place to fix it quickly, and I'm willing to take responsibility to fix it.

Reply:
This change was initially to unblock me on some issues with GPU compilation, but I'm happy to expand the scope if others are in agreement.

There's a handful of clang options we may want to keep to remove some regressions? Here's the set of flags that the `libc` project has right now:
```
-fPIC                                                                                                                                                                                                                   
-fno-semantic-interposition                                                                                                                                                                                             
-fvisibility-inlines-hidden                                                                                                                                                                                             
-Werror=date-time                                                                                                                                                                                                              
-Werror=unguarded-availability-new                                                                                                                                                                                             
-Wall -Wextra                                                                                                                                                                                                                  
-Wno-unused-parameter                                                                                                                                                                                                          
-Wwrite-strings -Wcast-qual                                                                                                                                                                                                    
-Wmissing-field-initializers                                                                                                                                                                                                   
-Wimplicit-fallthrough                                                                                                                                                                                                         
-Wcovered-switch-default                                                                                                                                                                                                       
-Wno-noexcept-type                                                                                                                                                                                                             
-Wnon-virtual-dtor                                                                                                                                                                                                             
-Wdelete-non-virtual-dtor                                                                                                                                                                                                      
-Wsuggest-override                                                                                                                                                                                                             
-Wstring-conversion                                                                                                                                                                                                            
-Wmisleading-indentation                                                                                                                                                                                                       
-Wctad-maybe-unsupported                                                                                                                                                                                                       
-fdiagnostics-color                                                                                                                                                                                                            
-ffunction-sections                                                                                                                                                                                                            
-fdata-sections
```

Do any of these look integral? We probably want to keep `-fPIC -ffunction-sections -fdata-sections` if nothing else. Warnings should probably also be somewhat consistent.

Reply:
> We could add the `include` in compiler-rt and in libc if we expect those to rely on it perhaps? For libc++, libc++abi and libunwind, I am fairly confident that if we remove it and our CI doesn't break, we're okay. It might break something, but if that's the case and if it's actively being used then we have pretty good systems in place to fix it quickly, and I'm willing to take responsibility to fix it.

The impetus for this was attempting to get math tests working in `libc` for the NVPTX GPU build. The automatic inclusion of `-Wl,` caused the linker to error as Nvidia's linker doesn't support pretty much anything. The other project I'd be interested in is `openmp`, which is a full runtime that statically links a lot of the LLVM libraries. I appreciate the help on this.

Reply:
FWIW, I think this is the right way to go. We'll handle the fallout as it shows up.

Reply:
@michaelrj-google Any thoughts about libc?

Basically, I want to avoid creating a haunted graveyard around these flags. We all agree they're undesirable and can be harmful, so let's do our best to remove them and see what needs to be fixed in addition. Let's not do the easy thing (this PR as-is) because it just sweeps future problems under the rug.

Reply:
That depends whether we just want to do the minimal thing now and clean up later or whether we want to try and do both at the same time. For example, I'd like to leverage native CMake option where possible, so instead of manually adding `-fPIC` to compiler flags, we should set `POSITION_INDEPENDENT_CODE` property on targets that actually need it. Feel free to make only the necessary minimum of changes you need and leave a `TODO`; we can follow up with more clean up changes later.